File: SharedContracts/ChargingSettings.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SharedContracts
{
    public class ChargingSettings
    {
        /// <summary>
        /// The configured charging profile
        /// 0 = off
        /// 1 = Access charging with battery priority
        /// 2 = Access charging with car priority
        /// 3 = Charging with PV power using battery power if needed
        /// 4 = Charging with PV power using grid power if needed
        /// 5 = Quick charging with 8 kW
        /// </summary>
        public int ChargingLevel { get; set; }

        /// <summary>
        /// The charging station that should be used first if both are available
        /// None-prefered station is used if the available power exceeds the minimum charging power of the first station (6A)
        /// </summary>
        public ChargingStation PreferedChargingStation { get; set; }

    };

    public enum ChargingStation
    {
        None = 0,
        Inside,
        Outside
    }
}

File: smarthome_app/lib/models/charging_settings.dart
enum ChargingStation { none, inside, outside }

class ChargingSettings {
  final int chargingLevel;
  final ChargingStation preferedChargingStation;
  final bool insideChargingEnabled;
  final bool outsideChargingEnabled;

  const ChargingSettings({
    this.chargingLevel = 0,
    this.preferedChargingStation = ChargingStation.none,
    this.insideChargingEnabled = true,
    this.outsideChargingEnabled = true,
  });

  factory ChargingSettings.fromJson(Map<String, dynamic> json) {
    final stationIndex = json['PreferedChargingStation'] as int? ?? 0;
    return ChargingSettings(
      chargingLevel: json['ChargingLevel'] as int? ?? 0,
      preferedChargingStation: stationIndex < ChargingStation.values.length
          ? ChargingStation.values[stationIndex]
          : ChargingStation.none,
      insideChargingEnabled: json['InsideChargingEnabled'] as bool? ?? true,
      outsideChargingEnabled: json['OutsideChargingEnabled'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
        'ChargingLevel': chargingLevel,
        'PreferedChargingStation': preferedChargingStation.index,
        'InsideChargingEnabled': insideChargingEnabled,
        'OutsideChargingEnabled': outsideChargingEnabled,
      };

  ChargingSettings copyWith({
    int? chargingLevel,
    ChargingStation? preferedChargingStation,
    bool? insideChargingEnabled,
    bool? outsideChargingEnabled,
  }) {
    return ChargingSettings(
      chargingLevel: chargingLevel ?? this.chargingLevel,
      preferedChargingStation:
          preferedChargingStation ?? this.preferedChargingStation,
      insideChargingEnabled:
          insideChargingEnabled ?? this.insideChargingEnabled,
      outsideChargingEnabled:
          outsideChargingEnabled ?? this.outsideChargingEnabled,
    );
  }
}
